pieforproviders/pieforproviders
# [585] Onboarding processing

Creates onboarding processor, onboarding downloader, and cleans up some associated/similar behavior for other processors, ensures same behavior for processor & controller when using Approval Amount Generator

# 💅🏼 What issue does this fix?
#585 - the scheduled job part will happen in Heroku

This is a big change because there was some wonky behavior that got surfaced as I was working on it.

Primary Functionality:
- adds new fields for Nebraska businesses, children/cases, and approvals
- adds NebraskaApprovalAmounts table to hold info about the approval amounts that come in on the onboarding spreadsheet to identify family fees
-  adds in an Onboarding Downloader to read in Onboarding Data from the AWS S3 bucket set in env vars for it
- adds a rake task to call the Onboarding Downloader so we can schedule it
- adds a processor to parse in all the information in the Onboarding CSV and create the appropriate businesses, cases, approvals, etc.
- adds archiving behavior so we can log and archive our failed imports to S3
- adds Nebraska Approval Amount Generator to read in from the CSV params and make an approval amount for each "grouped" approval amount (i.e. group of columns with the same Approval # in the header)

Minor Functionality:
- tweaks logging for attendance and dashboard imports because the labels were wrong
- removes old onboarding CSV logic from before Nebraska was the focus; new business info is changing that logic anyway and I wanted to clean it up to avoid confusion
- Reimplements License types as a module instead of an enum because we're learning tons more about licenses in the industry and it needs to be more flexible
- cleans up the "active_on_date" scope so it's consistent and works with our associations

## 🍂 Before You Submit


## 🛷 Deployment


## 🧵 Steps to set up locally

```
bundle exec rails db:migrate
```

## 🧳 Steps to test
- set an env var for all the S3 env vars in this PR - best if you set up an S3 bucket and get a key/access key ID for it yourself and test it out there because we're still working on IAM creds for our AWS Org
- put a fake csv in your S3 bucket, formatted like this: https://docs.google.com/spreadsheets/d/1fkxyQnSrchhzgjmTiS_wfuvtF8WVkdPz-qGxcXlrMSA/edit#gid=0 (make sure that the emails align with emails you have in your local app as existing, confirmed users)
- Call `bundle exec rails read_wonderschool_necc_dashboard_cases`
- You should see those cases in your DB along with appropriate and accurate approvals, businesses, etc.
- check your s3 archive bucket for failed results that will have been archived (if any)

## 🕯 Caveats, concerns
- There are a couple places where I've probably overdefined associations and "accepts nested attributes for" but I'm okay with cleaning those up later
